core/src/saros/concurrent/management/ConcurrentDocumentServer.java

```diff
@@ -53,38 +54,47 @@ public void userLeft(final User user) {
   public ConcurrentDocumentServer(final ISarosSession sarosSession) {
     this.sarosSession = sarosSession;
     this.server = new JupiterServer(sarosSession);
+
+    Consumer<SPath> deletedFileHandler =
+        resource -> {
+          LOG.debug("Resetting jupiter server for " + resource);
+          server.removePath(resource);
+        };
+
+    this.resourceActivityFilter = new ResourceActivityFilter(sarosSession, deletedFileHandler);
   }
 
   @Override
   public void start() {
+    if (!sarosSession.isHost()) {
```

Comment:
I know this is easy, but why do you not prevent this component being added to the container instead ?

Reply:
Because this causes issues with the PicoContainer. I see three possibilities to avoid adding it to the context:
1. Add `null` when not the host.
This causes the session container to throw an exception when initializing the context as `null` values are not accepted.
2. Don't add it at all when not the host.
This causes the session container to also throw an exception as it can no longer resolve all dependencies (even though the logical branches are never used, the logic requiring/using the server still exists on the client side).
3. Write a mock document server that is just a NOP for every method and add it when not the host.
Possible, but I didn't think it would make the situation much better. Also, this would require the introduction of the NOP class as well as an interface for the server and NOP-server to implement.

Or did I miss anything?

Reply:
I was interested in answer 2. Looks like the ActivityHandler is not decoupled well enough.

I would suggest either add an TODO here, stating that this behavior is wrong or you can try to add another CTOR to the ActivityHandler (without the Server) which would lead to null checks if the ConcurrentDocumentServer is available in the ActivityHandler.

Reply:
There already is a comment about this in the [`SarosCoreSessionContextFactory`](https://github.com/saros-project/saros/blob/master/core/src/saros/session/SarosCoreSessionContextFactory.java#L42). But I can add a comment for it here as well.

In general, I don't think this should be resolved in this PR. I will open an issue for it.

Reply:
I created the issue #719 for this. I also added a FIXME to the `start()` method referencing the issue.